lib/api/objectPutPart.js

```diff
@@ -322,7 +323,12 @@ function objectPutPart(authInfo, request, streamingV4Params, log,
                             error: err,
                             method: 'objectPutPart::metadata.putObjectMD',
                         });
-                        return next(err, destinationBucket);
+                        // Make a best effort single attempt to cleanup the
+                        // orphaned data while returning the original error from
+                        // the metadata layer.
+                        return dataDelete(dataGetInfo, log, () => {
```

Comment:
I would still keep this sync, as-in wait for callback. There were reasons before but given that batch delete is now in place we no longer should be adopting this pattern.

Reply:
The waterfall callback `next` is only called after the `dataDelete` function's callback is invoked. I've preserved the metadata failure error here.

Reply:
Thanks, I overlooked that.